File: Array/78. Subsets.js
`Given an integer array nums of unique elements, return all possible subsets (the power set).

The solution set must not contain duplicate subsets. Return the solution in any order.

 

Example 1:

Input: nums = [1,2,3]
Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
Example 2:

Input: nums = [0]
Output: [[],[0]]`;

const subsets = (nums) => {
  let tempDataStruct = [];
  const result = [];
  let idx = 0;
  const helper = (stack, index) => {
    result.push([...stack]);
    for (let i = index; i < nums.length; i++) {
      stack.push(nums[i]);
      helper(stack, i + 1);
      stack.pop();
    }
  };
  helper(tempDataStruct, idx);
  return result;
};

console.log(subsets([1, 2, 3]));

File: Array/78. Subsets.py

from typing import List
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result=[]
        self.helper(nums,0,[],result)
        return result

    def helper(self,nums:List[int],index:int,path:List[int],result:List[List[int]]) -> None:
        # base condition to add the path to the result
        result.append(path.copy())

        for i in range(index,len(nums)):
            path.append(nums[i])
            # recursively call the helper function with the updated path and index
            self.helper(nums,i+1,path,result)
            path.pop()


# Time complexity: O(2^n)
# Space complexity: O(2^n)
#
print(Solution().subsets([1,2,3])) # [[],[1],[1,2],[1,2,3],[1,3],[2],[2,3],[3]]
